yapapi/runner/__init__.py

```diff
@@ -438,30 +434,45 @@ async def fill_work_q():
         # Py38: find_offers_task.set_name('find_offers_task')
         try:
             task_fill_q = loop.create_task(fill_work_q())
-            services = {
+            get_done_task: Optional[asyncio.Task[Task[D, R]]] = None
+            services: Set[asyncio.Task[Any]] = {
                 find_offers_task,
                 task_fill_q,
                 loop.create_task(worker_starter()),
                 process_invoices_job,
             }
+
             while (
                 task_fill_q in services
                 or not work_queue.empty()
-                or tasks_processed["s"] > tasks_processed["c"]
+                or not done_queue.empty()
+                or tasks_started > tasks_completed
             ):
                 if datetime.now(timezone.utc) > self._expires:
                     raise TimeoutError(f"task timeout exceeded. timeout={self._conf.timeout}")
+
+                if not get_done_task:
+                    get_done_task = loop.create_task(done_queue.get())
```

Comment:
This (asyncio) task just removes a single completed (yapapi) task from done_queue.